src/schema/schema_list.c

```diff
@@ -9,115 +9,56 @@
 #include "wt_internal.h"
 
 /*
- * __schema_add_table --
- *	Add a table handle to the session's cache.
- */
-static int
-__schema_add_table(WT_SESSION_IMPL *session,
-    const char *name, size_t namelen, bool ok_incomplete, WT_TABLE **tablep)
-{
-	WT_DECL_RET;
-	WT_TABLE *table;
-	uint64_t bucket;
-
-	table = NULL;			/* -Wconditional-uninitialized */
-
-	/* Make sure the metadata is open before getting other locks. */
-	WT_RET(__wt_metadata_cursor(session, NULL));
-
-	WT_WITH_TABLE_READ_LOCK(session,
-	    ret = __wt_schema_open_table(
-	    session, name, namelen, ok_incomplete, &table));
-	WT_RET(ret);
-
-	bucket = table->name_hash % WT_HASH_ARRAY_SIZE;
-	TAILQ_INSERT_HEAD(&session->tables, table, q);
-	TAILQ_INSERT_HEAD(&session->tablehash[bucket], table, hashq);
-	*tablep = table;
-
-	return (0);
-}
-
-/*
- * __schema_find_table --
- *	Find the table handle for the named table in the session cache.
- */
-static int
-__schema_find_table(WT_SESSION_IMPL *session,
-    const char *name, size_t namelen, WT_TABLE **tablep)
-{
-	WT_TABLE *table;
-	const char *tablename;
-	uint64_t bucket;
-
-	bucket = __wt_hash_city64(name, namelen) % WT_HASH_ARRAY_SIZE;
-
-restart:
-	TAILQ_FOREACH(table, &session->tablehash[bucket], hashq) {
-		tablename = table->name;
-		(void)WT_PREFIX_SKIP(tablename, "table:");
-		if (WT_STRING_MATCH(tablename, name, namelen)) {
-			/*
-			 * Ignore stale tables.
-			 *
-			 * XXX: should be managed the same as btree handles,
-			 * with a local cache in each session and a shared list
-			 * in the connection.  There is still a race here
-			 * between checking the generation and opening the
-			 * first column group.
-			 */
-			if (table->schema_gen !=
-			    __wt_gen(session, WT_GEN_SCHEMA)) {
-				if (table->refcnt == 0) {
-					WT_RET(__wt_schema_remove_table(
-					    session, table));
-					goto restart;
-				}
-				continue;
-			}
-			*tablep = table;
-			return (0);
-		}
-	}
-
-	return (WT_NOTFOUND);
-}
-
-/*
  * __wt_schema_get_table --
  *	Get the table handle for the named table.
  */
 int
 __wt_schema_get_table(WT_SESSION_IMPL *session,
-    const char *name, size_t namelen, bool ok_incomplete, WT_TABLE **tablep)
+    const char *name, size_t namelen, bool ok_incomplete, uint32_t flags,
+    WT_TABLE **tablep)
 {
+	WT_DATA_HANDLE *saved_dhandle;
 	WT_DECL_RET;
+	WT_DECL_ITEM(namebuf);
 	WT_TABLE *table;
 
-	*tablep = table = NULL;
-	ret = __schema_find_table(session, name, namelen, &table);
-
-	if (ret == WT_NOTFOUND)
-		ret = __schema_add_table(
-		    session, name, namelen, ok_incomplete, &table);
+	saved_dhandle = session->dhandle;
 
-	if (ret == 0) {
-		++table->refcnt;
-		*tablep = table;
+	*tablep = NULL;
+	WT_RET(__wt_scr_alloc(session, namelen + 1, &namebuf));
+	WT_ERR(__wt_buf_fmt(
+	    session, namebuf, "table:%.*s", (int)namelen, name));
```

Comment:
In light of WT-3516, I think we should consider whether always allocating and copying a string here will add too much overhead to a common path.  Maybe we need another version of `__wt_schema_get_table` for the case where the caller has a `table:T` URI already...